File: kotlin_jetpack_compose/app/src/main/java/com/example/testapp/data/firebase/repository/UserRepositoryImpl.kt
package com.example.testapp.data.firebase.repository

import com.example.testapp.data.firebase.model.mapper.toSportasDomain
import com.example.testapp.data.firebase.network.service.FirebaseUserService
import com.example.testapp.domain.model.SportasUser
import com.example.testapp.domain.repository.UserRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class UserRepositoryFirebase(
    private val firebaseUserService: FirebaseUserService
) : UserRepository {

    override suspend fun getSportasById(id: String): SportasUser? =
        withContext(Dispatchers.IO) {
            try {
                val korisnik = firebaseUserService.getKorisnikById(id)
                korisnik?.toSportasDomain(id)
            } catch (e: Exception) {
                null
            }
        }

    override suspend fun getCurrentSportas(): SportasUser =
        withContext(Dispatchers.IO) {
            val uid = firebaseUserService.getCurrentUserId()
                ?: throw IllegalStateException("Korisnik nije prijavljen.")

            getSportasById(uid)
                ?: throw IllegalStateException("Profil sportaša nije pronađen.")
        }
}

File: flutter/test_app/lib/data/firebase/repository/user_repository_impl.dart
import 'package:test_app/data/firebase/model/user_mapper.dart';

import '../../../domain/model/sportas_user.dart';
import '../../../domain/repository/user_repository.dart';
import '../../firebase/network/service/user_service.dart';
import '../network/service/user_service.dart';

class FirebaseUserRepositoryImpl implements UserRepository {
  final FirebaseUserService firebaseUserService;

  FirebaseUserRepositoryImpl(this.firebaseUserService);

  @override
  Future<SportasUser?> getSportasUser(String id) async {
    try {
      final korisnik = await firebaseUserService.getKorisnikById(id);
      if (korisnik == null) return null;

      if (!korisnik.jeSportas) return null;

      return korisnik.toSportasDomain(id);
    } catch (_) {
      return null;
    }
  }

  Future<SportasUser> getCurrentSportas() async {
    final uid = firebaseUserService.getCurrentUserId();
    if (uid == null) {
      throw StateError('Korisnik nije prijavljen.');
    }

    final sportas = await getSportasUser(uid);
    if (sportas == null) {
      throw StateError('Profil sportaša nije pronađen.');
    }

    return sportas;
  }
}